feg/gateway/diameter/connection.go

```diff
@@ -155,16 +155,42 @@ func (c *Connection) getDiamConnection() (diam.Conn, *smpeer.Metadata, error) {
 		return nil, nil, errors.New("Could not obtain metadata from connection")
 	}
 	c.conn, c.metadata = conn, metadata
+	if cn, ok := conn.(diam.CloseNotifier); ok && cn != nil {
+		go c.connCloseNotify(cn.CloseNotify(), conn)
+	} else {
+		glog.Errorf("new diam conn (%T) %s -> %s is not CloseNotifier", conn, localAddr, c.server.Addr)
+	}
 	return conn, metadata, nil
 }
 
+func (c *Connection) connCloseNotify(cnc <-chan struct{}, conn diam.Conn) {
+	<-cnc // wait for close notifier
+	if glog.V(2) {
+		remoteAddress, localAddress := "nil", "nil"
+		if conn != nil {
+			if conn.LocalAddr() != nil {
+				localAddress = conn.LocalAddr().String()
+			}
+			if conn.LocalAddr() != nil {
+				remoteAddress = conn.RemoteAddr().String()
+			}
+		}
+		glog.Infof("notified of %s->%s connection closure", localAddress, remoteAddress)
+	}
+	if c != nil && conn != nil && c.destroyConnection(conn) {
+		// if connection was closed not by connection management functions, recover it
+		c.getDiamConnection()
```

Comment:
is it possible this fails again and then the connection is just lost again?

Meaning if for some reason there is a network outage and the server is not reacheable  `getDiamConnection`  will fail and the connection will be lost. 

Maybe we need some kind of infinite loop to keep retrying every n seconds?

Reply:
I think, if there is a prolonged network outage, we'll just fall back to active recovery - when the client tries to send out a request, getDiamConnection is called again and will attempt to reconnect. Putting infinite loop here could be a bit heavy, maybe a limited number of retries.

Reply:
Note `newConnection` suffers of this issue. It never tries to reconnect if it fails once. 
Maybe we could add a loop in that go routine and just call `newConnection` in this line  instead of `c.getDiamConnection()`

http://github.com/magma/magma/blob/master/feg/gateway/diameter/connection.go#L48-L48